### PR TITLE
[CET-912] Optimize group/system filters

### PR DESCRIPTION
**Pull request by nikhilunni**
## Change description

Turn group/system/etc filters from `O(#services * #groups)` to `O(#groups)` every time there's an update to one of the filters.

## Type of change

- [ ] Bug fix (fixes an issue)
- [ ] New feature (adds functionality)
- [ ] Improvement (improves codebase without changing functionality)

## Related issues

> Fix [#1]()

### Development

- [ ] Lint rules pass locally
- [ ] Application changes have been tested thoroughly
- [ ] Automated tests covering modified code pass

### Security

- [ ] Security impact of change has been considered
- [ ] Code follows company security practices and guidelines

### Code review

- [ ] Pull request has a descriptive title and context useful to a reviewer. Screenshots or screencasts are attached as necessary
- [ ] "Ready for review" label attached and reviewers assigned
- [ ] Changes have been reviewed by at least one other contributor
- [ ] Pull request linked to task tracker where applicable

### CODE REVIEW

File: src/components/Scorecards/ScorecardDetailsPage/ScorecardFilterCard/ScorecardFilterCard.tsx [hunk 1]
```diff
@@ -38,6 +38,7 @@ import {
   defaultGroupRefContext,
   defaultSystemRefContext,
 } from '../../../../utils/ComponentUtils';
+import {EntityFilterGroup} from "../../../../filters";
```
Comment (FoodProduct) [inline src/components/Scorecards/ScorecardDetailsPage/ScorecardFilterCard/ScorecardFilterCard.tsx:41]: nit:
```suggestion
import { EntityFilterGroup } from "../../../../filters";
```

File: src/utils/hooks.ts [hunk 2]
```diff
@@ -250,42 +250,61 @@ export function useFilters<T>(
...
+
+            return mapSoFar
+          }, {} as Record<string, Set<string>>);
+      })
+  }, [allFilterGroups, components]);
+
+  const filterGroups = useMemo(() => {
+    return components === undefined || groupPropertyLookup === undefined
+      ? undefined
+      : allFilterGroups.map((filterGroup, index) => {
```
Comment (FoodProduct) [inline src/utils/hooks.ts:280]: nit: this doesn't need to be a ternary 😝 
```suggestion
    if (components === undefined || groupPropertyLookup === undefined) {
      return;
    }
    return allFilterGroups.map((filterGroup, index) => {
```
```

File: src/utils/hooks.ts [hunk 3]
```diff
@@ -250,42 +250,61 @@ export function useFilters<T>(
...
+                  : groupProperty,
+                value: groupProperty,
               };
             },
-          };
-        });
+          ),
+          generatePredicate: (groupProperty: string) => {
+            return (t: T) => {
+              return groupPropertyLookup[index][groupProperty].has(stringifyAnyEntityRef(entityRef(t)))
```
Comment (FoodProduct) [inline src/utils/hooks.ts:301]: Is this right? if `groupPropertyLookup` is `Record<string, Set<string>>` then I expect it to look something like
```json
{
  "key": ["item1", "item2"]
}
```
This looks like it's trying to access the Record with a number (which will likely return `undefined`) and then attempt to access a string property off of that.

Comment (FoodProduct) [reply src/utils/hooks.ts:301]: Oops it's late, I missed the list nature of `groupPropertyNature` 🙈. I think this is OK...
```ts
const groupPropertyLookup: Record<string, Set<string>>[]
```

**Review by FoodProduct — APPROVED** [review]
I think it looks OK, but highly suggest tests for this since it's a bit complicated 🙏

File: src/components/Scorecards/ScorecardDetailsPage/ScorecardFilterCard/ScorecardFilterCard.tsx (hunk 1 above)
Comment (cristinabuenahora) [reply src/components/Scorecards/ScorecardDetailsPage/ScorecardFilterCard/ScorecardFilterCard.tsx:41]: I thought we did??

Comment (FoodProduct) [reply src/components/Scorecards/ScorecardDetailsPage/ScorecardFilterCard/ScorecardFilterCard.tsx:41]: I see that we have `eslint` set up, but apparently it doesn't care about this particular thing. I honestly never remember what is enforced by prettier vs eslint so not sure exactly why this didn't get caught without investigating.

**Pull request merged**